File: web-ui/backend/utils/validation.py
"""
Input validation utilities for backend
"""
import re
import html
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, validator


def sanitize_html(text: str) -> str:
    """
    Sanitize HTML to prevent XSS attacks
    
    Args:
        text: Input text
        
    Returns:
        Sanitized text
    """
    return html.escape(text)


def is_valid_command_input(text: str) -> tuple[bool, Optional[str]]:
    """
    Validate command input for dangerous patterns
    
    Args:
        text: Command input text
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    # Check for null bytes
    if '\0' in text:
        return False, "Input contains null bytes"
    
    # Check for excessive length
    if len(text) > 10000:
        return False, "Input is too long (max 10000 characters)"
    
    # Check for script injection attempts
    dangerous_patterns = [
        r'<script',
        r'javascript:',
        r'on\w+\s*=',  # Event handlers
        r'eval\s*\(',
        r'expression\s*\(',
    ]
    
    for pattern in dangerous_patterns:
        if re.search(pattern, text, re.IGNORECASE):
            return False, f"Input contains potentially dangerous pattern: {pattern}"
    
    return True, None


def is_valid_file_path(path: str) -> tuple[bool, Optional[str]]:
    """
    Validate file path
    
    Args:
        path: File path
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    # Check for null bytes
    if '\0' in path:
        return False, "Path contains null bytes"
    
    # Check for path traversal
    if '../' in path or '..\\'  in path:
        return False, "Path contains traversal attempts"
    
    # Check for invalid characters (Windows)
    invalid_chars = r'[<>:"|?*]'
    if re.search(invalid_chars, path):
        return False, "Path contains invalid characters"
    
    return True, None


def is_valid_username(username: str) -> tuple[bool, Optional[str]]:
    """
    Validate username format
    
    Args:
        username: Username
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not re.match(r'^[a-zA-Z0-9_-]{3,20}$', username):
        return False, "Username must be 3-20 characters and contain only letters, numbers, underscore, and hyphen"
    
    return True, None


def is_valid_email(email: str) -> tuple[bool, Optional[str]]:
    """
    Validate email format
    
    Args:
        email: Email address
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    email_pattern = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
    if not re.match(email_pattern, email):
        return False, "Invalid email format"
    
    return True, None


def is_strong_password(password: str) -> tuple[bool, Optional[str]]:
    """
    Validate password strength
    
    Args:
        password: Password
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if len(password) < 8:
        return False, "Password must be at least 8 characters"
    
    if not re.search(r'[A-Z]', password):
        return False, "Password must contain at least one uppercase letter"
    
    if not re.search(r'[a-z]', password):
        return False, "Password must contain at least one lowercase letter"
    
    if not re.search(r'[0-9]', password):
        return False, "Password must contain at least one number"
    
    return True, None


def sanitize_input(text: str) -> str:
    """
    Remove dangerous characters from input
    
    Args:
        text: Input text
        
    Returns:
        Sanitized text
    """
    # Remove null bytes
    text = text.replace('\0', '')
    
    # Remove control characters except newline and tab
    text = re.sub(r'[\x00-\x08\x0B-\x0C\x0E-\x1F\x7F]', '', text)
    
    return text


def validate_json_structure(data: Dict[str, Any], required_fields: List[str]) -> tuple[bool, Optional[str]]:
    """
    Validate JSON structure has required fields
    
    Args:
        data: JSON data
        required_fields: List of required field names
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    missing_fields = [field for field in required_fields if field not in data]
    
    if missing_fields:
        return False, f"Missing required fields: {', '.join(missing_fields)}"
    
    return True, None


def is_valid_port(port: int) -> tuple[bool, Optional[str]]:
    """
    Validate port number
    
    Args:
        port: Port number
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not isinstance(port, int):
        return False, "Port must be an integer"
    
    if port < 1 or port > 65535:
        return False, "Port must be between 1 and 65535"
    
    return True, None


def is_valid_ipv4(ip: str) -> tuple[bool, Optional[str]]:
    """
    Validate IPv4 address
    
    Args:
        ip: IP address
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    ipv4_pattern = r'^(\d{1,3}\.){3}\d{1,3}$'
    if not re.match(ipv4_pattern, ip):
        return False, "Invalid IPv4 format"
    
    parts = ip.split('.')
    for part in parts:
        num = int(part)
        if num < 0 or num > 255:
            return False, "IPv4 octets must be between 0 and 255"
    
    return True, None


class ValidationError(Exception):
    """Custom validation error"""
    pass


def validate_and_sanitize_command_input(text: str) -> str:
    """
    Validate and sanitize command input
    
    Args:
        text: Command input
        
    Returns:
        Sanitized text
        
    Raises:
        ValidationError: If validation fails
    """
    # Validate
    is_valid, error_msg = is_valid_command_input(text)
    if not is_valid:
        raise ValidationError(error_msg)
    
    # Sanitize
    return sanitize_input(text)


def validate_and_sanitize_file_path(path: str) -> str:
    """
    Validate and sanitize file path
    
    Args:
        path: File path
        
    Returns:
        Sanitized path
        
    Raises:
        ValidationError: If validation fails
    """
    # Validate
    is_valid, error_msg = is_valid_file_path(path)
    if not is_valid:
        raise ValidationError(error_msg)
    
    # Sanitize
    return sanitize_input(path)

File: web-ui/src/utils/validation.ts
/**
 * Input validation utilities
 */

/**
 * Sanitize HTML to prevent XSS attacks
 */
export function sanitizeHtml(input: string): string {
  const div = document.createElement('div');
  div.textContent = input;
  return div.innerHTML;
}

/**
 * Escape HTML special characters
 */
export function escapeHtml(input: string): string {
  const map: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#x27;',
    '/': '&#x2F;',
  };
  return input.replace(/[&<>"'/]/g, (char) => map[char]);
}

/**
 * Validate email format
 */
export function isValidEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Validate username (alphanumeric, underscore, hyphen, 3-20 chars)
 */
export function isValidUsername(username: string): boolean {
  const usernameRegex = /^[a-zA-Z0-9_-]{3,20}$/;
  return usernameRegex.test(username);
}

/**
 * Validate password strength
 * - At least 8 characters
 * - Contains at least one uppercase letter
 * - Contains at least one lowercase letter
 * - Contains at least one number
 */
export function isStrongPassword(password: string): boolean {
  if (password.length < 8) return false;
  if (!/[A-Z]/.test(password)) return false;
  if (!/[a-z]/.test(password)) return false;
  if (!/[0-9]/.test(password)) return false;
  return true;
}

/**
 * Get password strength level
 */
export function getPasswordStrength(password: string): 'weak' | 'medium' | 'strong' {
  let strength = 0;
  
  if (password.length >= 8) strength++;
  if (password.length >= 12) strength++;
  if (/[A-Z]/.test(password)) strength++;
  if (/[a-z]/.test(password)) strength++;
  if (/[0-9]/.test(password)) strength++;
  if (/[^A-Za-z0-9]/.test(password)) strength++;
  
  if (strength <= 2) return 'weak';
  if (strength <= 4) return 'medium';
  return 'strong';
}

/**
 * Validate URL format
 */
export function isValidUrl(url: string): boolean {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

/**
 * Validate IP address (IPv4)
 */
export function isValidIpv4(ip: string): boolean {
  const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
  if (!ipv4Regex.test(ip)) return false;
  
  const parts = ip.split('.');
  return parts.every(part => {
    const num = parseInt(part, 10);
    return num >= 0 && num <= 255;
  });
}

/**
 * Validate port number
 */
export function isValidPort(port: number | string): boolean {
  const portNum = typeof port === 'string' ? parseInt(port, 10) : port;
  return !isNaN(portNum) && portNum >= 1 && portNum <= 65535;
}

/**
 * Validate file path (basic check)
 */
export function isValidFilePath(path: string): boolean {
  // Check for null bytes
  if (path.includes('\0')) return false;
  
  // Check for path traversal attempts
  if (path.includes('../') || path.includes('..\\')) return false;
  
  // Check for invalid characters (Windows)
  const invalidChars = /[<>:"|?*]/;
  if (invalidChars.test(path)) return false;
  
  return true;
}

/**
 * Validate command input (check for dangerous patterns)
 */
export function isValidCommandInput(input: string): { valid: boolean; reason?: string } {
  // Check for null bytes
  if (input.includes('\0')) {
    return { valid: false, reason: 'Input contains null bytes' };
  }
  
  // Check for excessive length
  if (input.length > 10000) {
    return { valid: false, reason: 'Input is too long (max 10000 characters)' };
  }
  
  // Check for script injection attempts
  const dangerousPatterns = [
    /<script/i,
    /javascript:/i,
    /on\w+\s*=/i, // Event handlers like onclick=
    /eval\s*\(/i,
    /expression\s*\(/i,
  ];
  
  for (const pattern of dangerousPatterns) {
    if (pattern.test(input)) {
      return { valid: false, reason: 'Input contains potentially dangerous patterns' };
    }
  }
  
  return { valid: true };
}

/**
 * Validate JSON string
 */
export function isValidJson(str: string): boolean {
  try {
    JSON.parse(str);
    return true;
  } catch {
    return false;
  }
}

/**
 * Validate number range
 */
export function isInRange(value: number, min: number, max: number): boolean {
  return value >= min && value <= max;
}

/**
 * Validate string length
 */
export function isValidLength(str: string, min: number, max: number): boolean {
  return str.length >= min && str.length <= max;
}

/**
 * Validate required field
 */
export function isRequired(value: any): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === 'string' && value.trim() === '') return false;
  if (Array.isArray(value) && value.length === 0) return false;
  return true;
}

/**
 * Validate alphanumeric string
 */
export function isAlphanumeric(str: string): boolean {
  return /^[a-zA-Z0-9]+$/.test(str);
}

/**
 * Validate numeric string
 */
export function isNumeric(str: string): boolean {
  return /^\d+$/.test(str);
}

/**
 * Validate hexadecimal string
 */
export function isHexadecimal(str: string): boolean {
  return /^[0-9A-Fa-f]+$/.test(str);
}

/**
 * Remove dangerous characters from input
 */
export function sanitizeInput(input: string): string {
  // Remove null bytes
  let sanitized = input.replace(/\0/g, '');
  
  // Remove control characters except newline and tab
  sanitized = sanitized.replace(/[\x00-\x08\x0B-\x0C\x0E-\x1F\x7F]/g, '');
  
  return sanitized;
}

/**
 * Validate form data
 */
export interface ValidationRule {
  required?: boolean;
  minLength?: number;
  maxLength?: number;
  pattern?: RegExp;
  custom?: (value: any) => boolean;
  message?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: Record<string, string>;
}

export function validateForm(
  data: Record<string, any>,
  rules: Record<string, ValidationRule>
): ValidationResult {
  const errors: Record<string, string> = {};
  
  for (const [field, rule] of Object.entries(rules)) {
    const value = data[field];
    
    // Check required
    if (rule.required && !isRequired(value)) {
      errors[field] = rule.message || `${field} is required`;
      continue;
    }
    
    // Skip other validations if value is empty and not required
    if (!isRequired(value)) continue;
    
    // Check min length
    if (rule.minLength && typeof value === 'string' && value.length < rule.minLength) {
      errors[field] = rule.message || `${field} must be at least ${rule.minLength} characters`;
      continue;
    }
    
    // Check max length
    if (rule.maxLength && typeof value === 'string' && value.length > rule.maxLength) {
      errors[field] = rule.message || `${field} must be at most ${rule.maxLength} characters`;
      continue;
    }
    
    // Check pattern
    if (rule.pattern && typeof value === 'string' && !rule.pattern.test(value)) {
      errors[field] = rule.message || `${field} format is invalid`;
      continue;
    }
    
    // Check custom validation
    if (rule.custom && !rule.custom(value)) {
      errors[field] = rule.message || `${field} is invalid`;
      continue;
    }
  }
  
  return {
    valid: Object.keys(errors).length === 0,
    errors
  };
}
